ada_feeding/ada_feeding/idioms/pre_moveto_config.py

```diff
@@ -0,0 +1,226 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""
+This module defines the pre_moveto_config idiom, which returns a behavior that
+calls the ROS services that should be called before any MoveTo behavior.
+Specifically, it does the following:
+    1. Re-tare the force-torque sensor.
+        a. Toggle the Watchdog Listener off. This is necessary because re-taring
+           the force-torque sensor will briefly stop force-torque sensor readings,
+           which would cause the watchdog to fail.
+        b. Re-tare the force-torque sensor.
+        c. Toggle the Watchdog Listener on.
+    2. Set force-torque thresholds (so the ForceGateController will trip if
+       the force-torque sensor readings exceed the thresholds).
+"""
+
+# Standard imports
+import logging
+import operator
+from typing import Optional
+
+# Third-part_y imports
+import py_trees
+from py_trees.blackboard import Blackboard
+from rcl_interfaces.msg import Parameter, ParameterType, ParameterValue
+from rcl_interfaces.srv import SetParameters
+from std_srvs.srv import SetBool
+
+# Local imports
+from .retry_call_ros_service import retry_call_ros_service
+
+
+def set_parameter_response_all_success(
+    blackboard_value: SetParameters.Response,
+    _: SetParameters.Request,
+) -> bool:
+    """
+    Checks that all the parameters were successfully set.
+    """
+    return all(result.successful for result in blackboard_value.results)
+
+
+def pre_moveto_config(
+    name: str,
+    re_tare: bool = True,
+    toggle_watchdog_listener: bool = True,
+    f_mag: float = 0.0,
+    f_x: float = 0.0,
+    f_y: float = 0.0,
+    f_z: float = 0.0,
+    t_mag: float = 0.0,
+    t_x: float = 0.0,
+    t_y: float = 0.0,
+    t_z: float = 0.0,
+    logger: Optional[logging.Logger] = None,
+) -> py_trees.behaviour.Behaviour:
+    """
+    Returns a behavior that calls the ROS services that should be called before
+    any MoveTo behavior. Specifically, it does the following, in a Sequence:
+        1. Re-tare the force-torque sensor.
+            a. Toggle the Watchdog Listener off. This is necessary because re-taring
+               the force-torque sensor will briefly stop force-torque sensor readings,
+               which would cause the watchdog to fail.
+            b. Re-tare the force-torque sensor.
+            c. Toggle the Watchdog Listener on.
+        2. Set force-torque thresholds (so the ForceGateController will trip if
+           the force-torque sensor readings exceed the thresholds).
+
+    Parameters
+    ----------
+    name: The name to associate with this behavior.
+    re_tare: Whether to re-tare the force-torque sensor.
+    toggle_watchdog_listener: Whether to toggle the watchdog listener on and off.
+        In practice, if the watchdog listener is on, you should toggle it.
+    f_mag: The magnitude of the overall force threshold. No threshold if 0.0.
```

Comment:
A bit confused about these statements...is this saying that if f_mag=0.0 then it signifies no threshold?

Reply:
That is correct. The current default (all 0.0) basically keeps moving irrespective of the forces/torques. [This is how the ForceGateController is defined](https://github.com/personalrobotics/pr_ros_controllers/blob/f3c11af37a2c53e8de5f5ea39c723f1e2f031d27/force_gate_controller/src/force_gate_controller.cpp#L164).

I had them all default to 0 because [that is what the ForceGateController does](https://github.com/personalrobotics/pr_ros_controllers/blob/main/force_gate_controller/src/force_gate_controller_parameters.yaml), but I am realizing now that that is not a good idea for feeding. I'll change `f_mag` to default to 4, which [is the default when ada_moveit is launched](https://github.com/personalrobotics/ada_ros2/blob/e1229060469d3ae17c2ada28030cf9f78715a70d/ada_moveit/config/real_controllers.yaml#L36).